### ISSUE TITLE
Clouddriver reloading all the dynamic kubernetes accounts in every refresh even though there is no change of account information

### ISSUE THREAD

**Issue by kirangodishala** [spinnaker/spinnaker#4938]
### Issue Summary:
Clouddriver reloading all the dynamic kubernetes accounts in every refresh even though there is no change of account information

### Cloud Provider(s):
Kubernetes V2

### Environment:
Spinnaker 1.15.x, 1.16.x.

### Feature Area:
Clouddriver dynamic account feature

### Description:
With few hundreds of Kubernetes accounts configured through Dynamic account feature, the initial load itself is taking considerable minutes and after that even without any change in account information, we observe the same delay in every scheduled refresh. We expect loading only those accounts which are either added or modified.

Any fix to this issue is highly appreciated as we have been facing this issue for couple of months now.

### Steps to Reproduce:
"Dynamic account for kubernetes"
Add at least 800 Kubernetes accounts  in the remote(ex: git) clouddriver.yml file. Check for the log statements similar to the following. Note the timestamp which confirms high resource utilization for no change in account information.
2019-09-10 10:52:57.018  INFO 6 --- [reshScheduler-1] o.s.boot.SpringApplication               : Started application in 13.163 seconds (JVM running for 4416.644)
2019-09-10 11:00:43.311  INFO 6 --- [reshScheduler-1] k.v.c.KubernetesV2ProviderSynchronizable : No changes detected to V2 Kubernetes accounts. Skipping caching agent synchronization.

**Comment by ethanfrogers** [comment]
@kirangodishala I was under the impressions that this had been fixed in a previous commit but your comment seems to show that is wasn't. I'll check into this.

PR I believe should have fixed this - https://github.com/spinnaker/clouddriver/pull/3990

**Comment by kirangodishala** [comment]
@ethanfrogers, I am using this image: gcr.io/spinnaker-marketplace/**clouddriver:6.4.0-20190918202815**  which is just 5 days old, but I am seeing these logs which confirm the creation new temp file every time.
Here are the logs from the clouddriver:
_2019-09-24 04:55:52.583  INFO 1 --- [           main] o.s.c.c.s.e.NativeEnvironmentRepository  : Adding property source: file:/tmp/config-repo-2759101066525547105/clouddriver.yml
...
2019-09-24 05:04:34.419  INFO 1 --- [reshScheduler-1] o.s.c.c.s.e.NativeEnvironmentRepository  : Adding property source: file:/tmp/config-repo-8762382573774657466/clouddriver.yml
.....
2019-09-24 05:12:19.167  INFO 1 --- [reshScheduler-1] o.s.c.c.s.e.NativeEnvironmentRepository  : Adding property source: file:/tmp/config-repo-5295145829271279354/clouddriver.yml
.....
2019-09-24 05:20:28.378  INFO 1 --- [reshScheduler-1] o.s.c.c.s.e.NativeEnvironmentRepository  : Adding property source: file:/tmp/config-repo-3438968423780868431/clouddriver.yml_
![Screen Shot 2019-09-24 at 11 06 40 AM](https://user-images.githubusercontent.com/53332225/65483952-9f0a1700-debb-11e9-9fa5-3e033f429777.png)

**Comment by phantooom** [comment]
i have the same problem. when i change git config. git config has been downlaod. but my k8s account list did’t change.  @kirangodishala  did you reslove this problem? @ezimanyi  is it a bug? or my config mistike

```
spring:
  profiles:
    include: git
  cloud:
    config:
      server:
        git:
          uri: https://github.com/phantooom/cloud-config.git
          username: xiaorui.zou@gmail.com
          password: xxxxxxxx
          refresh-rate: 10
```

```
bash-4.4$ ls -l
total 32
drwx------    3 spinnake nogroup       4096 Oct 11 02:55 config-repo-5575538485973815976
drwx------    3 spinnake nogroup       4096 Oct 11 02:53 config-repo-733094432630290751
drwx------    3 spinnake nogroup       4096 Oct 11 03:24 config-repo-8436836329708476161
drwxr-xr-x    2 spinnake nogroup       4096 Oct 11 02:52 hsperfdata_spinnaker
-rw-r--r--    1 spinnake nogroup       5145 Oct 11 02:55 kubeconfig.yml
drwxr-xr-x    2 spinnake nogroup       4096 Oct 11 02:53 tomcat-docbase.6242963847963437906.7002
drwxr-xr-x    3 spinnake nogroup       4096 Oct 11 02:53 tomcat.3864702786157342963.7002
```

**Comment by RocketRaccoon** [comment]
Same issue to me. Has configuration both for clouddriver and igor. I see uploaded configs and logs but no changes applied in UI. Using spinnaker 1.16.3.

**Comment by scottfrederick** [comment]
One issue we identified with the Spring Cloud Config git integration is that the git repo is getting cloned to a new temp dir on each refresh. This is due to how Spring reloads beans on refresh. When using the git backend for Spring Cloud Config, you should explicitly set a `basedir` property so that the same directory is used on each refresh: 

```
spring:
  profiles:
    include: git
  cloud:
    config:
      server:
        git:
          uri: https://github.com/example/cloud-config.git
          basedir: /tmp/cloud-config
```

**Comment by kirangodishala** [comment]
> One issue we identified with the Spring Cloud Config git integration is that the git repo is getting cloned to a new temp dir on each refresh. This is due to how Spring reloads beans on refresh. When using the git backend for Spring Cloud Config, you should explicitly set a `basedir` property so that the same directory is used on each refresh:
> 
> ```
> spring:
>   profiles:
>     include: git
>   cloud:
>     config:
>       server:
>         git:
>           uri: https://github.com/example/cloud-config.git
>           basedir: /tmp/cloud-config
> ```
> 
> If you configure multiple git repo backends, each should have a unique `basedir`. We'll add this to the [external config docs](https://www.spinnaker.io/setup/configuration/).

Yeah, we found this "basedir" config too and thought it would fix the refresh issue but other than keeping the same directory and not updating the file under /tmp/config directory when there are no changes, it still  takes 20-30min to discover that there are no changes.

**Comment by jonathanheaden** [comment]
We are having a similar but not quite the same issue;
We have account information being served for two accounts by the config server, however if we make any changes they don't show up in the spinnaker UI. I set the `basedir` as above and can see that only one folder gets created on the clouddriver pod for the repo, and also that it does get the change if we make one.

Here are the steps taken:

1. Create clouddriver.yml in the git repo with 3 entries: default & 2 other accounts
1. Remove one of the accounts from the clouddriver.yml file
1. Observe change in the repo that is cloned to the clouddriver pod

The change is not then reflected in the spinnaker UI

**Comment by phantooom** [comment]
@jonathanheaden i have the same problem.

**Comment by kirangodishala** [comment]
@jonathanheaden, @phantooom -  It could be browser caching. Can you try opening in a different browser after the change?

**Comment by phantooom** [comment]
@kirangodishala  i did, i open a new chrome. but it did not change. than i restart clouddriver and open a new chrome. it works.

**Comment by phantooom** [comment]
```kubeconfigFile: configserver:kubeconfig.yml```  is it a key point?

**Comment by jonathanheaden** [comment]
I suspect that restarting clouddriver pod would also work for us, but as we are in proof of concept at the moment our configuration for clouddriver involves some manual steps so this is not currentl scalable. Also from my understanding it is not supposed to require a pod restart to pick up the config change

**Comment by kirangodishala** [comment]
Clouddriver profiling information is attached. I would request everyone to go through the documentation and provide your insights towards resolving this issue.

[Clouddriver_Profiling_Data.pdf](https://github.com/spinnaker/spinnaker/files/3767306/Clouddriver_Profiling_Data.pdf)
[Clouddriver_Profiling_SupportingData.zip](https://github.com/spinnaker/spinnaker/files/3767307/Clouddriver_Profiling_SupportingData.zip)

**Comment by ncknt** [comment]
Refreshing should work with `basedir` once https://github.com/spinnaker/kork/pull/404 is merged. I'd also set `spring.cloud.config.server.git.refreshRate` to `30` or more so the repo is not pulled for every `configserver` change.

**Comment by scottfrederick** [comment]
> once spinnaker/kork#404 is merged

Then kork should get auto-bumped in clouddriver, then clouddriver will get a new build.

**Comment by jonathanheaden** [comment]
What worked for us is that we switched to absolute file paths for the kubeconfig files (after copying these locally to the clouddriver pod) and then saw dynamic accounts working as expected. Previously we had been using `kubeconfigFile: configserver:kubeconfig.yml` to reference the kubeconfig file.

**Comment by scottfrederick** [comment]
> What worked for us is that we switched to absolute file paths for the kubeconfig files (after copying these locally to the clouddriver pod)

@jonathanheaden Can you expand on this? Are you copying the `kubeconfig` files to the clouddriver pod explicitly, or are you referring to the local copy that is being done by the `kork:config` code? Can you show an example of the absolute file path?

**Comment by ncknt** [comment]
I think @jonathanheaden may be referring to `basedir` (e.g. `/tmp/config` + hardcoded path (`/tmp/config/kubeconfig1.yml`). It's more efficient but the kubeconfig file can be changed by Clouddriver itself. I haven't checked if that actually triggers an account change.

**Comment by jonathanheaden** [comment]
in the `clouddriver-local.yml` file that the config server is using we reference the kubeconfig file as `kubeconfigFile: /home/spinnaker/serviceaccount-kubeconfig.yml` 
then we copy the file to `/home/spinnaker/serviceaccount-kubeconfig.yml` on the clouddriver pod.   
We use cloudbuild to place the kubeconfig files on the halyard pod as of now, so would look to re use that method to place them on the clouddriver pod if that is the approach we end up taking

**Comment by kirangodishala** [comment]
I think we are talking about two different issues. The issue I mentioned has nothing to do with KubeConfigFile. When the number of kubernetes accounts are as high as 800, clouddriver (in fact Spring Boot property binding process) goes on utilizing ~100% CPU for 20-30min.

I have created a small project to demonstrate this : https://github.com/kirangodishala/spring-cloud-config-server-client

K8sRestController is under RefreshScope and has K8sProperties autowired.

When the number of accounts are as high as 400, the /refresh endpoint upon triggering takes ~7min to respond back event if no changes are made the remote file.

**Comment by scottfrederick** [comment]
Several issues got piled onto this thread. I believe all of the issues have been addressed except: 

- the time taken for config load/refresh
- config changes appearing in the UI

The rest of this comment applies to the former, and the latter should be a new issue if people are still having problems. 

I've spent some time with the standalone example @kirangodishala shared in the comment above to understand what's going on with mapping the kubernetes account properties at startup and at refresh. 

**Config Server overhead**

> With few hundreds of Kubernetes accounts configured through Dynamic account feature, the initial load itself is taking considerable minutes...

I ran the standalone example as provided (with a Config Server and Config Client). I also modified the client app to remove the dependency on `spring-cloud-starter-config` and copy the [`clouddriver.yml` from the config repo](https://github.com/kirangodishala/cloud-config/blob/master/clouddriver.yml) to an `application.yml` file loaded directly by the client app. The result is that the client app takes several minutes to start when the 400 kubernetes accounts are loaded from config server or from a local `application.yml`. The Config Server/Config Client does not add any measurable time to what the Boot property mapping is doing to load the accounts at startup. 

**Config refresh**

> ... and after that even without any change in account information, we observe the same delay in every scheduled refresh. We expect loading only those accounts which are either added or modified.

What does change with Config Server/Config Client vs just loading from `application.yml` is of course the refresh. Instead of Boot only loading properties at app startup, it also loads properties with each refresh.  

On refresh, the config is requested again from Config Server and Config Server just returns what it has without any knowledge of whether anything has changed. 

Once Config Client gets the config it will compile a list of any property keys that have changed since the last refresh and report the list of changed keys to the bean refresh code, but it only does this after Boot has mapped the new YAML from Config Server to a new `@ConfigurationProperties` bean so it can compare. It doesn't keep a previous copy of YAML around to compare against. So the time taken here is still in Boot, not in refreshing any beans. 

**Size matters**

In the [Slack thread](https://spinnakerteam.slack.com/archives/C748G8U9J/p1571640591001900) that preceded this issue, @ethanfrogers reported app startup and refresh times much faster than what @kirangodishala experienced with up to 500 accounts. I'm not sure of all the factors that contributed to this difference, but I can confirm that the number of fields included in the kubernetes account config makes a big difference in the time it takes Boot to map the properties. The accounts in Kiran's generated [test config](https://github.com/kirangodishala/cloud-config/blob/master/clouddriver.yml) include every field supported by the `@ConfigurationProperties` bean, including many empty arrays and an empty map. This is the config that takes several minutes to hydrate. 

Reducing the config for each account to the following fields cut the time by around 30%. 

```
    - name: k8s-v2-acc-1
      providerVersion: V2
      namespaces:
        - ns11
        - ns12
        - ns13
        - ns14
        - ns15
      omitKinds:
        - clusterRoleBinding
        - controllerRevision
        - customResourceDefinition
        - role
        - roleBinding
        - persistentVolume
        - apiService
        - validatingWebhookConfiguration
        - podPreset
        - mutatingWebhookConfiguration
        - podSecurityPolicydone
      kubeconfigFile: /home/spinnaker/.hal/default/staging/dependencies/1573052617-config
```

Removing one of the remaining arrays cut the time to around 30 seconds: 

```
    - name: k8s-v2-acc-1
      providerVersion: V2
      namespaces:
        - ns11
        - ns12
        - ns13
        - ns14
        - ns15
      kubeconfigFile: /home/spinnaker/.hal/default/staging/dependencies/1573052617-config
```

Removing everything but the bare minimum cut the time to around 7 seconds: 

```
    - name: k8s-v2-acc-1
      providerVersion: V2
      kubeconfigFile: /home/spinnaker/.hal/default/staging/dependencies/1573052617-config
```

**Next steps**

I think this all comes down to Boot's property loading taking a long time to process the quantity of data in this type of kubernetes config. We can dig deeper in to this involve the Spring Boot team if necessary. 

We've tied the config refresh interval to the provider caching agent interval (with a default of 1 minute) pending feedback on automatic refresh. We should add a separate property to set the refresh interval, and also allow automatic refresh to be disabled so that it can be triggered manually to better handle cases where the refresh might take a while.

**Comment by clanesf** [comment]
Commenting because this is important to me. Haven't had time to track this since about 1.15, but back then if we didn't specify namespaces and custom resource definitions it could slow clouddriver dramatically. In particular the credentials call. Here is an issue I filed about it. https://github.com/spinnaker/spinnaker/issues/4310

 I believe @ezimanyi traced it to a cache inside of each clouddriver pod that was storing namespaces strangely and this may be corrected as of 1.17.

**Comment by scottfrederick** [comment]
> I think this all comes down to Boot's property loading taking a long time to process the quantity of data in this type of kubernetes config. We can dig deeper in to this involve the Spring Boot team if necessary.

I further narrowed this down to the performance of Spring Boot when loading very large YAML files containing array structures. Most of the time spent mapping the properties YAML to the `@ConfigurationProperties` bean is being spent analyzing the flattened properties to group array properties together and determine size. As part of this investigation with a simple test app, I happened to compare the performance of Spring Boot 2.1.7 (the version currently used by Spinnaker services) to Spring Boot 2.2.1 (the latest version). It turns out that Boot 2.2.1 is significantly faster in this area, potentially by a factor of 10. 

The next step test upgrading clouddriver to Boot 2.2.1 and testing the effect that has on clouddriver startup with hundreds of `kubernetes` accounts. I'm working on that now.

**Comment by robzienert** [comment]
> The next step test upgrading clouddriver to Boot 2.2.1

Just an FYI: Mind you that upgrading a service from one boot to another migrates _all_ services, since we manage dependencies via kork's `spinnaker-dependencies` module, and kork is auto-bumped across all services. :)

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by clanesf** [comment]
@spinnakerbot remove-label stale

This is still very much an issue for us, and with the Spring Boot upgrade being delayed it becomes an even bigger issue.

**Comment by RocketRaccoon** [comment]
Any updates on issue?

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by spinnakerbot** [comment]
This issue is tagged as 'stale' and hasn't been updated  in 45 days, so we are tagging it as 'to-be-closed'. It will be closed in 45 days unless updates are made. If you want to remove this label, comment:

> @spinnakerbot remove-label to-be-closed

**Comment by spinnakerbot** [comment]
This issue is tagged as 'to-be-closed' and hasn't been updated  in 45 days, so we are closing it. You can always reopen this issue if needed.
